karapace/kafka_rest_apis/__init__.py

```diff
@@ -279,24 +279,31 @@ def get_topic_config(self, topic: str) -> dict:
             return self.admin_client.get_topic_config(topic)
 
     def cluster_metadata(self, topics: Optional[List[str]] = None) -> dict:
-        if self._metadata_birth is None or time.time() - self._metadata_birth > self.metadata_max_age:
-            self._cluster_metadata = None
-        if not self._cluster_metadata:
-            self._metadata_birth = time.time()
-            with self.admin_lock:
-                try:
-                    self._cluster_metadata = self.admin_client.cluster_metadata(topics)
-                except NodeNotReadyError:
-                    self.log.exception("Could not refresh cluster metadata")
-                    self.r(
-                        body={
-                            "message": "Kafka node not ready",
-                            "code": RESTErrorCodes.HTTP_INTERNAL_SERVER_ERROR.value,
-                        },
-                        content_type="application/json",
-                        status=HTTPStatus.INTERNAL_SERVER_ERROR,
-                    )
-        return copy.deepcopy(self._cluster_metadata)
+        with self.admin_lock:
```

Comment:
Aiohttp is providing the thread. in a case where full metadata is requested and processing takes some time and a subsequent request is done it is possible to have deadlock as `threading.Lock` would be already held by the same thread. I would propose to use `asyncio.Lock`, this lock must be created in the eventloop thread.

Reply:
There was already `_async_provider_lock = asyncio.Lock()` so it was easy to verify it is created in the eventloop thread.